src/unix/battery.c

```diff
@@ -0,0 +1,252 @@
+                                  RTLD_LAZY | RTLD_LOCAL);
+
+  if (iokit_framekwork_handle == NULL || core_foundation_handle == NULL)
+    goto out;
+
+  *(void **)(&pCFRelease) = dlsym(core_foundation_handle, "CFRelease");
+  *(void **)(&pCFStringCreateWithCString) =
+      dlsym(core_foundation_handle, "CFStringCreateWithCString");
+  *(void **)(&pCFArrayGetCount) =
+      dlsym(core_foundation_handle, "CFArrayGetCount");
+  *(void **)(&pCFArrayGetValueAtIndex) =
+      dlsym(core_foundation_handle, "CFArrayGetValueAtIndex");
+  *(void **)(&pCFNumberGetValue) =
+      dlsym(core_foundation_handle, "CFNumberGetValue");
+  *(void **)(&pCFBooleanGetValue) =
+      dlsym(core_foundation_handle, "CFBooleanGetValue");
+  *(void **)(&pCFDictionaryGetValue) =
+      dlsym(core_foundation_handle, "CFDictionaryGetValue");
+  *(void **)(&pIOPSCopyPowerSourcesInfo) =
+      dlsym(iokit_framekwork_handle, "IOPSCopyPowerSourcesInfo");
+  *(void **)(&pIOPSCopyPowerSourcesList) =
+      dlsym(iokit_framekwork_handle, "IOPSCopyPowerSourcesList");
+  *(void **)(&pIOPSGetPowerSourceDescription) =
+      dlsym(iokit_framekwork_handle, "IOPSGetPowerSourceDescription");
+
+  if (pCFRelease == NULL ||
+      pCFStringCreateWithCString == NULL ||
+      pCFArrayGetCount == NULL ||
+      pCFArrayGetValueAtIndex == NULL ||
+      pCFNumberGetValue == NULL ||
+      pCFDictionaryGetValue == NULL ||
+      pIOPSCopyPowerSourcesInfo == NULL ||
+      pIOPSCopyPowerSourcesList == NULL ||
+      pIOPSGetPowerSourceDescription == NULL) {
+    goto out;
+  }
+
+#define S(s) pCFStringCreateWithCString(NULL, (s), kCFStringEncodingUTF8)
+
+  blob = pIOPSCopyPowerSourcesInfo();
+  if (blob == NULL)
+    goto out;
+
+  sources = pIOPSCopyPowerSourcesList(blob);
+  if (sources == NULL)
+    goto out;
+
+  // If there are no sources, we're on a desktop or server, nothin' to do
+  if (pCFArrayGetCount(sources) == 0) {
+    err = UV_EPERM;
+    goto out;
+  }
+
+  powerSource =
+    pIOPSGetPowerSourceDescription(blob, pCFArrayGetValueAtIndex(sources, 0));
+  if (powerSource == NULL)
+    goto out;
+
+  err = 0;
+
+  // kIOPSCurrentCapacityKey
+  currentCapacity = pCFDictionaryGetValue(powerSource, S("Current Capacity"));
+  if (currentCapacity != NULL) {
+    pCFNumberGetValue(currentCapacity, pkCFNumberIntType, &holder);
+    info->level = holder;
+  }
+
+
+  // Retrieve is_charging status
+  // kIOPSIsChargingKey
+  isChargingValue = pCFDictionaryGetValue(powerSource, S("Is Charging"));
+  if (isChargingValue != NULL) {
+    is_charging = pCFBooleanGetValue(isChargingValue);
+    info->is_charging = is_charging;
+  }
+
+  // kIOPSTimeToFullChargeKey and kIOPSTimeToEmptyKey
+  timeRemainingValue =
+    pCFDictionaryGetValue(powerSource,
+        is_charging ? S("Time to Full Charge") : S("Time to Empty"));
+  if (timeRemainingValue != NULL) {
+    pCFNumberGetValue(timeRemainingValue, pkCFNumberIntType, &holder);
```

Comment:
Should maybe be called `current_capacity_value` or something like that? Unclear what `holder` refers to.

Reply:
I just created **one** variable to hold/store all values, I just rewrite that holder during the whole process, I will create all the required variables for this :)

Reply:
It's okay to have a single variable that's used repeatedly but we usually give those a single-letter (or very short) name, like `v`.